contracts/smart-account/modules/BatchedSessionRouterModule.sol

```diff
@@ -0,0 +1,190 @@
+     * If something goes wrong, reverts
+     * @param userOp User Operation to be validated.
+     * @param userOpHash Hash of the User Operation to be validated.
+     * @return SIG_VALIDATION_FAILED or packed validation result.
+     */
+    function validateUserOp(
+        UserOperation calldata userOp,
+        bytes32 userOpHash
+    ) external virtual returns (uint256) {
+        // check this is a proper method call
+        require(
+            bytes4(userOp.callData[0:4]) == EXECUTE_BATCH_OPTIMIZED_SELECTOR ||
+                bytes4(userOp.callData[0:4]) == EXECUTE_BATCH_SELECTOR,
+            "SR Invalid Selector"
+        );
+
+        (bytes memory moduleSignature, ) = abi.decode(
+            userOp.signature,
+            (bytes, address)
+        );
+
+        /*
+        struct SessionCallData {
+            uint48 validUntil;
+            uint48 validAfter;
+            address sessionValidationModule;
+            bytes sessionKeyData;
+            bytes32[] merkleProof;
+            bytes callSpecificData;
+        }
+        */
+
+        // parse the signature to get the array of required parameters
+        (
+            address sessionKeyManager,
+            uint48[] memory validUntil,
+            uint48[] memory validAfter,
+            address[] memory sessionValidationModule,
+            bytes[] memory sessionKeyData,
+            bytes32[][] memory merkleProof,
+            bytes[] memory callSpecificData,
+            bytes memory sessionKeySignature
+        ) = abi.decode(
+                moduleSignature,
+                (
+                    address,
+                    uint48[],
+                    uint48[],
+                    address[],
+                    bytes[],
+                    bytes32[][],
+                    bytes[],
+                    bytes
+                )
+            );
+
+        // can I use struct instead so don't have to check lengths
+
+        // check lengths of arrays
+        require(
+            validUntil.length == validAfter.length,
+            "SR Invalid data provided"
+        );
+        require(
+            validUntil.length == sessionValidationModule.length,
+            "SR Invalid data provided"
+        );
+        require(
+            validUntil.length == sessionKeyData.length,
+            "SR Invalid data provided"
+        );
+        require(
+            validUntil.length == merkleProof.length,
+            "SR Invalid data provided"
+        );
+
+        address recovered = ECDSA.recover(
+            ECDSA.toEthSignedMessageHash(
+                keccak256(abi.encodePacked(userOpHash, sessionKeyManager))
+            ),
+            sessionKeySignature
+        );
+
+        uint48 earliestValidUntil = type(uint48).max;
+        uint48 latestValidAfter;
+
+        // iterate over batched operations
+        for (uint i; i < sessionValidationModule.length; ) {
```

Comment:
Given that `sessionValidationModule` is a memory array, it's length should be cached to save gas

Reply:
my experiments showed that this caching doesn't always saves gas
https://twitter.com/filmakarov/status/1642067067644911618?s=20

Reply:
Ah yes I've read your thread before, for the experiment with a memory based array you did see a reduction in gas afaik https://twitter.com/filmakarov/status/1642067079443501056?t=T-iKe55pPjbj36UnjblTOQ&s=19

Reply:
You're right. fixed.